### ISSUE TITLE
Encrypt cookies with AES-GCM

### ISSUE THREAD

**Issue by zzak** [sinatra/sinatra#1300]
Recently, Rails added AEAD encrypted cookies: rails/rails#28132

We currently encrypt using SHA1: https://github.com/sinatra/sinatra/blob/1b0edc0aeaaf4839cadfcec1b21da86e6af1d4c0/rack-protection/lib/rack/protection/base.rb#L111

/cc @jkowens @mikeycgto

**Comment by mwpastore** [comment]
We should at least use SHA2 out of the box.

**Comment by zzak** [comment]
I'm sorry I didn't think of this until now :disappointed:

**Comment by mikeycgto** [comment]
I'd be happy to look into this and submit a PR to upgrade the encryption.
With Rails, most of the complexity with my changes was around supporting
seamless upgrades from the now legacy encryption to this new scheme. Is
that something Sinatra would be interested in supporting?

http://michaeljcoyne.me

**Comment by mikeycgto** [comment]
Started to look more into this and I'm little confused as to where session cookies come from for Sinatra. Do they come from rack/session/cookie code base or do they solely come from the above mentioned file?

Either way, both are just doing a HMAC for cookie security. I think introducing AES encryption would certainly be a worthwhile addition to Sinatra (and thus introduce confidential to the system). Should I be looking to create a PR against the rack session code base or against the above mentioned code base?

Would love to contribute this feature, so please advice on what the best course of action would be. Thanks!!

**Comment by mikeycgto** [comment]
Thanks for the feedback. Will look into creating a PR sometime next week (as I am away in Europe at the moment). Will be great to have Sinatra and other rack apps benefit from more modern encryption!

**Comment by mikeycgto** [comment]
Put together a PR for `Rack::Session::Cookie`. Any and all feedback or questions is welcomed!

As far as `Rack::Protection::SessionHijacking` goes, the use the `encrypt` method is no longer needed. Since the underlying session is encrypted and signed, we can just rely on the security of that. Using AES for the encrypting the session data now introduces confidentiality, thus making the `SessionHijacking` module even more useful. A theoretical attack cannot view the session data and would not know what to forge the stored User-Agent and Language headers to.

I can make a PR to update this module.

**Comment by zzak** [comment]
@mikeycgto Does that depend on your PR to Rack getting merged and released?

**Comment by mikeycgto** [comment]
Yes, my [changes](https://github.com/sinatra/sinatra/compare/master...mikeycgto:session-hijack-changes) to rack/protection/session_hijacking would require my PR for Rack to be merged before that can be merged. I made a branch for it on my sinatra fork but have yet to open a PR.

Maybe we move my changes to the cookie middleware into an independent middleware that sinatra can use directly? Getting such a major change accepted into Rack may take a long while.

**Comment by zzak** [comment]
@mikeycgto Yeah, I'm :+1: to rolling this out into a shim or something once it's merged.

We will probably want to support the next major rack release, whenever that is, I guess rack 3 will be a thing. But until then we should still support your encrypted cookies

**Comment by mikeycgto** [comment]
Sounds good. I can take my changes to rack's cookie middleware and integrate them directly into sinatra for the short term. Will put together a PR for this soon.

**Comment by darkogj** [comment]
Rails already does this from version 5.2:

As of Rails 5.2 encrypted cookies and sessions are protected using AES GCM encryption. This form of encryption is a type of Authenticated Encryption and couples authentication and encryption in single step while also producing shorter ciphertexts as compared to other algorithms previously used. The key for cookies encrypted with AES GCM are derived using a salt value defined by the config.action_dispatch.authenticated_encrypted_cookie_salt configuration value.

Prior to this version, encrypted cookies were secured using AES in CBC mode with HMAC using SHA1 for authentication. The keys for this type of encryption and for HMAC verification were derived via the salts defined by config.action_dispatch.encrypted_cookie_salt and config.action_dispatch.encrypted_signed_cookie_salt respectively.

**Comment by mjc-gh** [comment]
There is now some movement on Rack encryption sessions here: https://github.com/rack/rack/pull/1177

Once this is merged, I would like to look into supporting encrypted cookies on Sinatra.

**Comment by jkowens** [comment]
Thanks @mjc-gh!

**Comment by ioquatix** [comment]
Thanks everyone for this ongoing effort.

We've recently moved the session handling code from `rack` to `rack-session` gem. This is to simplify versioning and make it easier to implement improvements to session handling independently of Rack as a web server interface.

We would love to have some attention to this issue over in `rack-session`. Specifically, I merged @mjc-gh's proposal into `rack-session` but would love an extra set of eyes (or several) before we release it properly - and perhaps even a working implementation from here?

**Comment by jkowens** [comment]
@mjc-gh I know this started over 5 years ago...but just wanted to see if you would be interested in moving forward with this now that your update with the shim was merged into master and the `rack-session` gem has been released with your updates?
